### ISSUE TITLE
cog-execute! is mangling BindLinks

### ISSUE THREAD

**Issue by linas** [opencog/atomspace#705]
Spin-off from bug #704 -- cog-execute, when given a BindLink, manges the bindlink. Example:

```
(use-modules (opencog))
(use-modules (opencog exec))
(use-modules (opencog query))

(define z-dsn (DefinedSchemaNode "OpenPsi: Energy-action-rule-Default"))
(DefineLink
   z-dsn
   (BindLink
      (VariableList
         (TypedVariableLink
            (VariableNode "Demand" (stv 0.96491325 0))
            (TypeNode "ConceptNode")
         )
      )
      (AndLink
         (DefinedPredicateNode "OpenPsi: demand-pattern")
         (EqualLink
            (VariableNode "Demand" (stv 0.96491325 0))
            (ConceptNode "OpenPsi: Energy" (stv 0.66960984 0.99999982))
         )
      )
      (ExecutionOutputLink
         (GroundedSchemaNode "scm: foo")
         (ListLink
            (VariableNode "Demand" (stv 0.96491325 0))
         )
      )
   )
)

(define (foo node) node)
```

Then  `(cog-execute! z-dsn)`  returns
```
(BindLink
   (VariableList
      (TypedVariableLink
         (VariableNode "Demand")
         (TypeNode "ConceptNode")
      )
   )
   (AndLink
      (DefinedPredicateNode "OpenPsi: demand-pattern")
      (EqualLink
         (VariableNode "Demand" )
         (ConceptNode "OpenPsi: Energy")
      )
   )
   (VariableNode "Demand" )
)
```
which is more or less junk.   This is related to bug #686

**Comment by linas** [comment]
easier said than done. Doing this promptly elicits infinite recursion in the two unit tests that check for it.  It also sends the forward chainer into infinite recursion, as well.

**Comment by linas** [comment]
pull request #713 fixes this.  After loading the above, and the other code from #704, the following output is generated:
```
(cog-execute! z-dsn)
$11 = (SetLink
   (ConceptNode "OpenPsi: Energy" (stv 0.66960984 0.99999982))
)
```

which I beleive is correct, given the lack of a no-exec link type.

**Comment by linas** [comment]
Fixed in pull req #713
